Preprocessor symbols for nullable reference types: Define a single nullable context.

BillWagner opened the pull request:
Fixes #1088 

Alternative PR to #1108

This branch builds on the work in #1108. This is an alternative approach that defines a single `nullable` context. The nullable context has two properties: `annotations` and `warnings` that can be enabled or disabled separately.

BillWagner committed:
Port text from #700

The first commit ports in the text from #700 substantially without change. The only editorial change is to move the clause on "Nullable directives" in lexical-structure before the clause on Pragma directives.

In addition, it fixes build warnings

BillWagner committed:
Remove link (which will need to get added)

Remove an Xref to a section that's not in this PR.

BillWagner committed:
Define a single nullable context

Alternative PR to #1108.

This branch builds on the work in #1108. This is an alternative approach that defines a single `nullable` context. The nullable context has two properties: `annotations` and `warnings` that can be enabled or disabled separately.

jskeet reviewed (approved):
Yup, personally I like this better than having multiple separate notional contexts.

Nigel-Ecma commented on standard/lexical-structure.md:
```diff
@@ -1507,6 +1509,40 @@ A `#line hidden` directive has no effect on the compilation unit and line number
+A nullable directive sets the denoted nullable context for subsequent lines of code, until another nullable directive overrides it, or until the end of the source code is reached. The effect of each form of nullable directive is, as follows:
+
+- `#nullable disable`: Sets the nullable context to “disabled”.
+- `#nullable enable`: Sets the nullable context to “enabled”.
```
Needs to say both annotations and warnings are enabled.

The previous bullet could also say it disables both, though that is not so important if the  options are:

- null context disabled, or
- null context enabled with annotations and/or warnings

jskeet replied:
Personally I'd prefer the approach of saying "enables both annotations and warnings".

jskeet replied:
In meeting: yes, make each of these specify the effect on both flags.

Nigel-Ecma commented on standard/lexical-structure.md:
```diff
@@ -1507,6 +1509,40 @@ A `#line hidden` directive has no effect on the compilation unit and line number
+- `#nullable enable warnings`: Enables nullable warnings. Nullable annotations are unaffected.
+- `#nullable restore warnings`: Restores nullable warnings to the state specified by the external mechanism, if any. Nullable annotations are unaffected.
+
+Disabling the nullable context when it is already disabled has no effect. Likewise, enabling the nullable context when it is already enabled has no effect.
```
`#nullable enable` has an effect if only one of annotations and warnings are currently enabled as it enables the other.

jskeet replied:
In meeting: no need for this.

Nigel-Ecma reviewed (changes requested):
A couple of language fixes/clarifications required so it describes the same overall semantics as #1108.

Nigel-Ecma commented:
> Yup, personally I like this better than having multiple separate notional contexts.

I probably agree (yes I'm hedging ;-)) – If a file starts with an enable/disable and never changes it makes little difference. However I can see that when nullable state is enabled & disabled at different points in a file having one context might make that semantic description simpler.

jskeet commented:
Example of something that might be hard to understand:

```csharp
#nullable disable
string x = null;
string y = "";
#nullable enable
Console.WriteLine(x.ToString());
Console.WriteLine(y.ToString());
```

Should that give a warning for `x.ToString()`?

#1124 states: "However, the disabled context means the compiler does not track null state and does not issue nullable related warnings" - so that would suggest no warning.

However, (with .NET 8 SDK) it issues a warning:

warning CS8602: Dereference of a possibly null reference.

jskeet reviewed (approved):
Approved but with changes as described.

BillWagner committed:
Update per meeting comments in review

This commit addresses all online feedback.

BillWagner committed:
Respond to feedback from June meeting

I listened to the June meeting. I updated a bit of text, and added Jon's example  from the conversation.

BillWagner commented:
I updated everything from the last meeting (both comments, and from the in-meeting discussion). This should be ready to merge @jskeet 

We can discuss at the meeting, or merge before.

jskeet commented:
> We can discuss at the meeting, or merge before.

Let's get a final review at the meeting just in case. It would be handy if everyone could review beforehand though.

BillWagner committed:
Update lexical-structure.md

Nigel-Ecma commented on standard/lexical-structure.md:
```diff
@@ -1026,7 +1026,7 @@ right_shift_assignment
 ### 6.5.1 General
 
-The pre-processing directives provide the ability to conditionally skip sections of compilation units, to report error and warning conditions, and to delineate distinct regions of source code.
+The pre-processing directives provide the ability to conditionally skip sections of compilation units, to report error and warning conditions, to delineate distinct regions of source code, and to set the nullable context.
```
Reading this now “nullable context” caused a pause. We have a null analysis, which may be enabled or disabled; and certain annotations (like adding a `?` to a type) etc. which both provide input to that analysis and which are treated differently depending on whether the analysis is being done. Is this a “nullable context”, a ”null analysis context”, something else?

BillWagner replied:
Tracked in the epic.

Nigel-Ecma replied:
Re: [epic comment](https://github.com/dotnet/csharpstandard/issues/1085#issuecomment-2221397275) – yes “is being done” should be worded ”is enabled/disabled”. Which unfortunately highlights something we are stuck with – the analysis is always being done but the pragmas enable/disable it… C’est la vie

Nigel-Ecma commented on standard/lexical-structure.md:
````diff
@@ -1507,6 +1509,56 @@ A `#line hidden` directive has no effect on the compilation unit and line number
+    ;
+```
+
+A nullable directive sets the available flags for subsequent lines of code, until another nullable directive overrides it, or until the end of the source code is reached. The nullable context contains two flags: *annotations* and *warnings*. The effect of each form of nullable directive is, as follows:
````
“until the end of the compilation unit is reached”? (or *compilation_unit*)

BillWagner replied:
```suggestion
A nullable directive sets the available flags for subsequent lines of code, until another nullable directive overrides it, or until the end of the *compilation _unit* is reached. The nullable context contains two flags: *annotations* and *warnings*. The effect of each form of nullable directive is, as follows:
```

Nigel-Ecma commented on standard/lexical-structure.md:
```diff
@@ -1507,6 +1509,56 @@ A `#line hidden` directive has no effect on the compilation unit and line number
+- `#nullable enable warnings`: Enables the nullable warnings flag. The nullable annotations flag is unaffected.
+- `#nullable restore warnings`: Restores the nullable warnings flag to the state specified by the external mechanism, if any. The nullable annotations flag is unaffected.
+
+The nullable state of expressions is tracked at all times. The state of the annotation flag determines the initial null state of a variable declaration. Warnings are only issued when the warnings flag is enabled.
```
“The nullable state of expressions…” vs. “…initial null state of a variable…”
Change the former to “null”?
“is tracked” -> ”shall be tracked” – that’s one for @RexJaeschke to determine, but the tense seems wrong to me and maybe the Standardese.

gafter replied:
The state of the annotation flag *and* the presence or absence of a nullable annotation `?`.

BillWagner replied:
```suggestion
The nullable state of expressions is tracked at all times. The state of the annotation flag and the presence or absence of a nullable annotation, `?`, determines the initial null state of a variable declaration. Warnings are only issued when the warnings flag is enabled.
```

Nigel-Ecma commented on standard/lexical-structure.md:
````diff
@@ -1507,6 +1509,56 @@ A `#line hidden` directive has no effect on the compilation unit and line number
+> Console.WriteLine(y.Length);
+> ```
+>
+> produces a compile-time warning (“dereference of a possible null reference”). The nullable state of `x` is tracked everywhere. A warning is issued when the warnings flag is enabled.
````
I would replace:

> (“dereference of a possible null reference”)

with

> (as `x` is `null`)

as the former reads like it is specifying the error text, including “possible”, even though we know its non-normative.

Of course this also states *non normatively* that there will be a warning, we is also not currently required…

BillWagner replied:
```suggestion
> produces a compile-time warning (“as `x` is `null`”). The nullable state of `x` is tracked everywhere. A warning is issued when the warnings flag is enabled.
```

Nigel-Ecma reviewed (dismissed):
A few language issues.

I also do not think we should approve/merge this, it is one of a set and as the others are developed we might need to circle around to this or any of them. So we should leave it as "done for now", a well known Github state ;-)

BillWagner committed:
Apply suggestions from code review

BillWagner committed:
Update standard/lexical-structure.md

jskeet commented:
> I also do not think we should approve/merge this, it is one of a set and as the others are developed we might need to circle around to this or any of them. So we should leave it as "done for now", a well known Github state ;-)

This was discussed unfortunately in Nigel's absence due to connectivity issues. We agreed that it would be better to avoid having multiple PRs hanging around and getting stale - the spec is "always inconsistent" in a draft form, as it doesn't accurately describe any specific version of C#.

If Nigel wishes to convince us otherwise, we could always revert the commit and open another PR :)

We will dismiss Nigel's review in order to merge now.
